geps/gep-1911.md

```diff
@@ -0,0 +1,199 @@
+# GEP-1911: Backend Protocol Selection
+
+* Issue: [#1911](https://github.com/kubernetes-sigs/gateway-api/issues/1911)
+* Status: Provisional
+
+(See status definitions [here](overview.md#status).)
+
+## TLDR
+
+Not all implementations support automatic protocol selection. Even in some cases protocols are disabled without an explicit opt-in (eg. websockets). Thus application developers need the ability to specify the protocol(s) that their application supports.
+
+## Goals
+
+- Support protocols that can have a Gateway '*Route' resource as a frontend
+- Consistent way of declaring backend protocols using a meta-resource
+- Support backends with multiple protocols on the same port (ie. h2c/http1)
+- Extensible - additional protocol specific fields should be easy to add in the future
+
+## Non-Goals
+
+- Backend TLS (covered in [GEP-1897](https://github.com/kubernetes-sigs/gateway-api/issues/1897)
+
+## Introduction
+
+## API
+
+We propose adding a new meta-resource [GEP-713](geps/gep-713.md) called `BackendProtocol`.
+
+This allows end-users to specify a list of ports and a list of corresponding protocols that that single
+port supports.
+
+### Go API
+
+```go
+type BackendProtocol struct {
+  // Standard type metadata
+  metav1.TypeMeta `json:",inline"`
+
+  // Standard object's metadata
+  metav1.ObjectMeta `json:"metadata,omitempty"`
+
+  // Spec defines the behaviour of this meta-resource
+  Spec BackendProtocolSpec `json:"spec,omitempty"`
+}
+
+type BackendProtocolSpec struct {
+  // Specifies the API resource who's protocols are being
+  // explicitly defined
+  TargetRef LocalObjectReference `json:"targetRef"`
+
+  // Ports contains a list of protocols for a specific ports
+  // When multiple protocols reference the same port the 
+  // entries earlier in the list should have higher priority
+  //
+  // ie. attempt to use h2c otherwise fallback on http1
+  //  ports: 
+  //   - port: 80
+  //     http2: 
+  //       encrypted: false # enable h2c
+  //   - port: 80
+  //     http1: {}          # We also listen to http1 on port 80
+  //
+  Ports []BackendPort `json:"ports,omitempty"`
+}
+
+type BackendPort struct {
+  // Number or name of the port who's protocol is being defined
+  Port intstr.IntOrString
+
+  PortProtocol `json:",inline"`
+}
+
+type PortProtocol struct {
+  HTTP2     *HTTP2PortProtocol    `json:"http2"`
+  HTTP1     *HTTP1PortProtocol    `json:"http1"`
+  GRPC      *GRPCPortProtocol     `json:"grpc"`
+  WebSocket *WebsocketPortProtocol`json:"websocket"`
+}
+
+type HTTP2PortProtocol struct {
+  Cleartext bool // h2c when set to true
+
+  // Future knobs can go here
+}
+
+type HTTPPortProtocol struct {
+  // Future knobs can go here
+}
+type GRPCPortProtocol struct {
+  // Future knobs can go here
+}
+type WebsocketPortProtocol struct {
+  // Future knobs can go here
+}
+```
+
+### YAML API
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1alpha1
+kind: BackendProtocol
```

Comment:
The service is k8s core/v1 has already include ports and protocol,, why not use  it